formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonDecoder.kt

```diff
@@ -74,12 +76,30 @@ internal open class StreamingJsonDecoder(
         }
     }
 
+    override fun <T> decodeSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        deserializer: DeserializationStrategy<T>,
+        previousValue: T?
+    ): T {
+        // TODO kludge for Map
+        val value = super.decodeSerializableElement(descriptor, index, deserializer, previousValue)
+        if (mode == WriteMode.MAP && index == 0) {
+            lexer.path.updateCurrentMapKey(value)
+        }
+        return value
+    }
+
     override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
-        return when (mode) {
+        val index = when (mode) {
             WriteMode.OBJ -> decodeObjectIndex(descriptor)
             WriteMode.MAP -> decodeMapIndex()
             else -> decodeListIndex() // Both for LIST and default polymorphic
         }
+        if (mode != WriteMode.MAP) {
+            lexer.path.updateCurrentIndex(index)
```

Comment:
You can move this to `.also {}` in `WriteMode.OBJ` and `else ->`. Not sure if this would save any time though.